docs/architecture/adr-0023-text-array-object-acrs.md

```diff
@@ -0,0 +1,166 @@
+
+When listing actions or permissions by subject type and id, which will potentially require going into each string of the text array, we could use postgres array functions such as `unnest()` combined with `like` to match string patterns.
+
+### Migration
+
+Migrations could be a bit tricky and probably could take some time to run. Adding a new field and updating this field in all first-class objects' tables would likely require a novel approach of creating a temp table with the new column, writing to a new permanent table, and dropping the previous table. Further research on it will be in a future issue card.
+
+Note that the `acrs` field will have an empty text array as the default value.
+
+### Examples for datasources
+
+This section shows some initial attempts for adapting to the ACR text array solution. These are for initial tests only and don't represent final implementation details. Datasources could be a simple case, but such adaptations can be applied to other first-class objects in the similar way.
+
+#### Data models
+
+```scala
+@JsonCodec
+case class Datasource(
+  id: UUID,
+  createdAt: java.sql.Timestamp,
+  createdBy: String,
+  modifiedAt: java.sql.Timestamp,
+  modifiedBy: String,
+  owner: String,
+  name: String,
+  visibility: Visibility,
+  composites: Json,
+  extras: Json,
+  bands: Json,
+  licenseName: Option[String],
+  acrs: List[String]
+)
+
+@JsonCodec
+case class ObjectAccessControlRuleCreate (
+  subjectType: SubjectType,
+  subjectIdO: Option[String],
+  actionType: ActionType
+) {
+  val subjectId: String = subjectIdO match {
+    case Some(subjectId) => subjectId
+    case _ => ""
+  }
+
+  def toAccessControlRuleString: String =
+    s"${subjectType.toString};${subjectId};${actionType.toString}"
+}
+```
+
+#### CRUD
+
+```scala
+def getAcrsByObjectId(object: Datasource): Fragment =
+  fr"SELECT acrs FROM datasources where id = ${datasource.id};"
+
+def updatePermissions(objectAcrCreateList: List[ObjectAccessControlRuleCreate], datasourceId: UUID, user: User): ConnectionIO[Int] = {
+    val now = new Timestamp((new java.util.Date()).getTime())
+    val objectAcrs = objectAcrCreateList.map(_.toAccessControlRuleString)
+    val updateQuery =
+      fr"UPDATE" ++ this.tableF ++ fr"SET" ++
+      fr"""
+        modified_at = ${now},
+        modified_by = ${user.id},
+        acrs = ${objectAcrs}
+      where id = ${datasourceId}
+      """
+    updateQuery.update.run
+}
+```
+
+#### First-class objects query filtered by ACRs
+
+```scala
+def exampleListQuery(user: User, actionType: ActionType, tableF: Fragment): Fragment = {
+  val baseF: Fragment = fr"SELECT * FROM" ++ tableF ++ "WHERE"
+  val ownedF: Fragment = fr"owner = ${user.id}"
+  val publicF: Fragment = fr"visibility = 'PUBLIC'"
+  val sharedPlatF: Fragment = fr"ARRAY[${"ALL;;" + actionType.toString}]"
+  val sharedUserF: Fragment = fr"ARRAY[${"USER;" + user.id + ";" + actionType.toString}]"
+  val inheritedF: Fragment = fr"""
```

Comment:
I'm wondering if instead of using a subquery here, it would be better practice for us to get a set of user group roles up front with a separate query.

Mostly I'm thinking about how easy it would be to extend some of the practices here if we assumed the function `exampleListQuery` either took a function (`user => UserGroupRoles`) or a list of `UserGroupRoles` which would allow us to rely on caching or parsing user group roles from the JWT. I'm wary of future proofing too much, but it also seems less magical in a way to avoid the SQL around array formation.

Reply:
Good point! I will put these as comments under that section for future development references.